pkgs/development/ocaml-modules/promise_jsoo/default.nix

```diff
@@ -0,0 +1,32 @@
+{ lib, fetchurl, buildDunePackage, ocaml, js_of_ocaml, ppxlib, js_of_ocaml-ppx, gen_js_api, ojs }:
+
+let
+  owner = "mnxn";
+in
+buildDunePackage rec {
+  pname = "promise_jsoo";
+  version = "0.3.1";
+  minimumOCamlVersion = "4.02";
+  useDune2 = lib.versionAtLeast ocaml.version "4.08";
+
+  src = fetchurl {
+    url = "https://github.com/${owner}/${pname}/releases/download/v${version}/${pname}-v${version}.tbz";
+    sha256 = "00pjnsbv0yv3hhxbbl8dsljgr95kjgi9w8j1x46gjyxg9zayrxzl";
+  };
+
+  buildInputs = [
+    js_of_ocaml
+    ppxlib
+    js_of_ocaml-ppx
+    gen_js_api
+  ];
+  propagatedBuildInputs = [ ojs ];
```

Comment:
Are you sure that `js_of_ocaml` need not be also propagated?

Reply:
Actually, `ojs` too doesn't need to be required. I made a mistake in my flake's buildInputs.

Reply:
I think that both `ojs` and `js_of_ocaml` must be added to `propagatedBuildInputs`. Did you try to use this package?

Reply:
Yes. I am using that package here - https://github.com/jayesh-bhoot/pin-unpin-tab/blob/master/flake.nix

I am able to `dune build` and test the add-on.

The way I reason about this is:

`ojs` is a runtime dependency of `gen_js_api`. So, whichever package depends on `gen_js_api` should need `ojs` while building the package (i.e., `gen_js_api`'s runtime). So `ojs` should be available while building this package.

Now, X listed under `propagatedBuildInputs` in derivation A implies that if derivation B lists A under its `buildInputs`, then B will have X available automatically. So B would only have to specify A in its own `buildInputs`, and not explicitly keep track of X as well.

So, inside A.nix: `propagatedBuildInputs = [ X ];`.
Inside B.nix: `buildInputs = [ A ];`.
Now, X is available to B too.

With this in mind, `gen_js_api` derivation (https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/ocaml-modules/gen_js_api/default.nix) already lists `ojs` under its `propagatedBuildInputs`. So `promise_jsoo` which depends on `gen_js_api` need not depend explicitly on `ojs` too.

Of course, I am still learning all this. Let me know if I am wrong.

Reply:
This is right. But users of `promise_jsoo` may not depend on `gen_js_api`; nonetheless, they do depend on `ojs`.

Reply:
Understood. `promise_jsoo` should work standalone.